File: common/display/src/main/scala/uk/ac/wellcome/display/models/DisplayLocation.scala
package uk.ac.wellcome.display.models

import io.circe.generic.extras.JsonKey
import io.swagger.v3.oas.annotations.media.Schema
import uk.ac.wellcome.models.work.internal.{DigitalResourceFormat, Location}

@Schema(
  name = "Location",
  description = "A location that provides access to an item",
  discriminatorProperty = "type",
  allOf = Array(
    classOf[Location.ClosedStores],
    classOf[Location.OpenShelves],
    classOf[Location.DigitalResource])
)
sealed trait DisplayLocation {
  val label: String
  val ontologyType: String
}

object DisplayLocation {
  def apply(location: Location): DisplayLocation = location match {
    case location: Location.ClosedStores =>
      DisplayClosedStores(location)
    case location: Location.OpenShelves => DisplayOpenShelves(location)
    case location: Location.DigitalResource =>
      DisplayDigitalResource(location)
  }

  case class DisplayClosedStores(accessConditions: List[DisplayAccessCondition],
                                 label: String = "Closed stores",
                                 @JsonKey("type") ontologyType: String =
                                   "ClosedStores")
      extends DisplayLocation

  object DisplayClosedStores {
    def apply(location: Location.ClosedStores): DisplayClosedStores =
      DisplayClosedStores(
        accessConditions =
          location.accessConditions.map(DisplayAccessCondition(_)))
  }
  case class DisplayOpenShelves(accessConditions: List[DisplayAccessCondition],
                                shelfmark: String,
                                shelfLocation: String,
                                label: String = "Open shelves",
                                @JsonKey("type") ontologyType: String =
                                  "OpenShelves")
      extends DisplayLocation

  object DisplayOpenShelves {
    def apply(location: Location.OpenShelves): DisplayOpenShelves =
      DisplayOpenShelves(
        accessConditions =
          location.accessConditions.map(DisplayAccessCondition(_)),
        shelfmark = location.shelfmark,
        shelfLocation = location.shelfLocation)
  }

  case class DisplayDigitalResource(
    accessConditions: List[DisplayAccessCondition],
    url: String,
    license: Option[DisplayLicense],
    credit: Option[String],
    format: Option[DisplayDigitalResourceFormat],
    label: String = "Online",
    @JsonKey("type") ontologyType: String = "DigitalLocation")
      extends DisplayLocation

  object DisplayDigitalResource {
    def apply(location: Location.DigitalResource): DisplayDigitalResource =
      DisplayDigitalResource(
        accessConditions =
          location.accessConditions.map(DisplayAccessCondition(_)),
        url = location.url,
        license = location.license.map(DisplayLicense(_)),
        credit = location.credit,
        format = location.format.map(DisplayDigitalResourceFormat(_))
      )
  }
}

case class DisplayDigitalResourceFormat(
  label: String,
  @JsonKey("type") @Schema(name = "type") ontologyType: String)
object DisplayDigitalResourceFormat {
  def apply(digitalResourceFormat: DigitalResourceFormat)
    : DisplayDigitalResourceFormat = digitalResourceFormat match {
    case DigitalResourceFormat.IIIFPresentation =>
      DisplayDigitalResourceFormat("IIIF presentation", "IIIFPresentation")
    case DigitalResourceFormat.IIIFImage =>
      DisplayDigitalResourceFormat("IIIF image", "IIIFImage")
  }
}

File: catalogue_graph/src/elasticsearch_transformers/display_location.py
from models.indexable_work import (
    DisplayDigitalLocation,
    DisplayIdLabel,
    DisplayLocation,
    DisplayPhysicalLocation,
)

from .display_access_condition import get_display_access_condition
from .display_license import get_display_license

LOCATION_LABEL_MAPPING = {
    "closed-stores": "Closed stores",
    "open-shelves": "Open shelves",
    "on-exhibition": "On exhibition",
    "on-order": "On order",
    "iiif-presentation": "IIIF Presentation API",
    "iiif-image": "IIIF Image API",
    "thumbnail-image": "Thumbnail image",
    "online-resource": "Online resource",
}

DIGITAL_LOCATIONS = {
    "iiif-presentation",
    "iiif-image",
    "thumbnail-image",
    "online-resource",
}


def get_display_location_type(location_type_id: str) -> DisplayIdLabel:
    return DisplayIdLabel(
        id=location_type_id,
        label=LOCATION_LABEL_MAPPING[location_type_id],
        type="LocationType",
    )


def get_display_location(
    raw_location,
) -> DisplayDigitalLocation | DisplayPhysicalLocation:
    location_type = raw_location["locationType"]["id"]
    license_id = raw_location.get("license", {}).get("id")
    is_digital = location_type in DIGITAL_LOCATIONS

    location = DisplayLocation(
        locationType=get_display_location_type(location_type),
        license=get_display_license(license_id) if license_id else None,
        accessConditions=[
            get_display_access_condition(c) for c in raw_location["accessConditions"]
        ],
    )

    if is_digital:
        return DisplayDigitalLocation(
            **location.dict(),
            url=raw_location["url"],
            credit=raw_location.get("credit"),
            linkText=raw_location.get("linkText"),
        )

    return DisplayPhysicalLocation(
        **location.dict(),
        label=raw_location["label"],
        shelfmark=raw_location.get("shelfmark"),
    )
